### PR TITLE
JAMES-3028 Bucket name resolver S3 blob store

### PR DESCRIPTION
**Pull request by Arsnael**
Having some issues with some unit tests with the blobstore S3 regarding the bucket prefix. If someone fancies taking a look, I disabled the corresponding tests in the code...

### CODE REVIEW

File: server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3BlobStoreDAO.java [hunk 1]
```diff
@@ -111,6 +114,11 @@
             .maxPendingAcquireUnbounded()
             .sizeUnbounded()
             .fifo();
+
+        bucketNameResolver = BucketNameResolver.builder()
```
Comment (chibenwa) [inline server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3BlobStoreDAO.java:118]: Maybe we can inject the all bucketname resolver stuff?

File: server/blob/blob-s3/src/test/java/org/apache/james/blob/objectstorage/aws/S3DeDuplicationBlobStoreTest.java [hunk 2]
```diff
@@ -37,18 +37,24 @@
...
+            .build();
+
+        s3BlobStoreDAO = new S3BlobStoreDAO(s3Configuration);
+
         testee = BlobStoreFactory.builder()
             .blobStoreDAO(s3BlobStoreDAO)
             .blobIdFactory(new HashBlobId.Factory())
             .defaultBucketName()
-            .passthrough();
+            .deduplication();
```
Comment (chibenwa) [inline server/blob/blob-s3/src/test/java/org/apache/james/blob/objectstorage/aws/S3DeDuplicationBlobStoreTest.java:57]: Were we playing dedup test suite against a passthrough blobstore by mistake?

Could it be the route cause of your troubles?

File: server/blob/blob-s3/src/test/java/org/apache/james/blob/objectstorage/aws/S3PrefixAndNamespaceTest.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    @Override
+    public BlobId.Factory blobIdFactory() {
+        return new HashBlobId.Factory();
+    }
+
+    @Override
+    @Test
+    @Disabled("JAMES-3028: No exception thrown in this case...")
+    public void readBytesStreamShouldThrowWhenBucketDoesNotExist() {
```
Comment (chibenwa) [inline server/blob/blob-s3/src/test/java/org/apache/james/blob/objectstorage/aws/S3PrefixAndNamespaceTest.java:93]: Does it returns something? Did you debug the actual bucket that had been used to communicate with S3 ?

**Review by chibenwa — COMMENTED** [review]
Good luck...

File: server/blob/blob-s3/src/test/java/org/apache/james/blob/objectstorage/aws/S3DeDuplicationBlobStoreTest.java (hunk 2 above)
Comment (Arsnael) [reply server/blob/blob-s3/src/test/java/org/apache/james/blob/objectstorage/aws/S3DeDuplicationBlobStoreTest.java:57]: No I found the root cause, nothing to do with this. That's just little fix, was testimng a passthrough in dedup tests and dedup in passthrough tests. Technically it doesn't change except matching correctly the test suite name

File: server/blob/blob-s3/src/test/java/org/apache/james/blob/objectstorage/aws/S3PrefixAndNamespaceTest.java (hunk 3 above)
Comment (Arsnael) [reply server/blob/blob-s3/src/test/java/org/apache/james/blob/objectstorage/aws/S3PrefixAndNamespaceTest.java:93]: Fix coming up :)

File: server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3BlobStoreDAO.java (hunk 1 above)
Comment (Arsnael) [reply server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3BlobStoreDAO.java:118]: I thought here that: 
* we only use bucket name resolver in S3BlobStoreDAO
* the bucket name resolver is being built with namespace + bucket prefix defined in conf
* I think what we needed until now was aws s3 auth conf + region conf... but if we had the bucket prefix and namespace for the resolver, then S3BlobStore needs all params from S3BlobStoreConfiguration. 

So I thought injecting directly the all conf and doing the stuff I need in the constructor of S3BlobStore was more simple. Allowed to get rid of the region and awsS3Auth Guice bidings in S3BlobStoreModule (as we don't seem to use them anywhere else). The integration tests are not breaking locally

**Comment by Arsnael** [comment]
The problem was that some public function reuse some other public functions in the S3BlobStoreDAO, so was doing sometimes two bucket name resolutions, which created some issues in some cases

**Comment by Arsnael** [comment]
However I don't have an idea how to check with an integration test that the bucket prefix is taken into account, as we only use the resolver on the S3BlobStoreDAO level and we randomly generate the bucket prefix in AwsS3Extension

File: server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3BlobStoreDAO.java (hunk 1 above)
Comment (chibenwa) [reply server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3BlobStoreDAO.java:118]: Let's keep this dumb simple then ;-)

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
No we also use the prefix policy for super trash, the only feature that leverage buckets (buckets are used to implement retention)

**Review by LanKhuat — APPROVED** [review]

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**